client/components/happychat/timeline.jsx

```diff
@@ -0,0 +1,169 @@
+import assign from 'lodash/assign';
+import { connect } from 'react-redux';
+
+/**
+ *	Internal dependencies
+ */
+import {
+	first,
+	when,
+	forEach,
+	propExists
+} from './functional';
+import autoscroll from './autoscroll';
+import AgentW from 'components/happychat/agent-w';
+import scrollbleed from './scrollbleed';
+import { translate } from 'i18n-calypso';
+import { getCurrentUser } from 'state/current-user/selectors';
+import {
+	getHappychatConnectionStatus,
+	getHappychatTimeline
+} from 'state/happychat/selectors';
+
+const debug = require( 'debug' )( 'calypso:happychat:timeline' );
+
+const linksNotEmpty = ( { links } ) => ! isEmpty( links );
+
+const messageParagraph = ( { message, key } ) => <p key={ key }>{ message }</p>;
+
+/*
+ * Given a message and array of links contained within that message, returns the message
+ * with clickable links inside of it.
+ */
+const messageWithLinks = ( { message, key, links } ) => {
+	const children = links.reduce( ( { parts, last }, [ url, startIndex, length ] ) => {
+		if ( last < startIndex ) {
+			parts = parts.concat( <span key={ parts.length }>{ message.slice( last, startIndex ) }</span> );
+		}
+		parts = parts.concat( <a key={ parts.length } href={ url } rel="noopener noreferrer" target="_blank">{ url }</a> );
+		return { parts, last: startIndex + length };
+	}, { parts: [], last: 0 } );
+
+	if ( children.last < message.length ) {
+		children.parts = children.parts.concat( <span key="last">{ message.slice( children.last ) }</span> );
+	}
+
+	return <p key={ key }>{ children.parts }</p>;
+};
+
+/*
+ * If a message event has a message with links in it, return a component with clickable links.
+ * Otherwise just return a single paragraph with the text.
+ */
+const messageText = when( linksNotEmpty, messageWithLinks, messageParagraph );
+const messageAvatar = when( propExists( 'meta.image' ), ( { meta } ) => <img alt={ meta.nick } src={ meta.image } /> );
+
+/*
+ * Group messages based on user so when any user sends multiple messages they will be grouped
+ * within the same message bubble until it reaches a message from a different user.
+ */
+const renderGroupedMessages = ( { item, isCurrentUser }, index ) => {
+	const [ event, ... rest ] = item;
+	const userAvatar = messageAvatar( { meta: event } );
+	return (
+		<div className={ classnames( 'happychat__timeline-message', { 'user-message': isCurrentUser } ) } key={ event.id || index }>
+			<div className="happychat__message-text">
+				{ messageText( {
+					message: event.message,
+					nick: event.nick,
+					key: event.id,
+					links: event.links
+				} ) }
+				{ rest.map( ( { message, id: key, links } ) => messageText( { message, key, links } ) ) }
+			</div>
+			<div className="happychat__message-meta">
+				<div className="happychat__message-avatar">
+				{ isCurrentUser ? userAvatar : <AgentW /> }
+				</div>
+			</div>
+		</div>
+	);
+};
+
+const itemTypeIs = type => ( { item: [ firstItem ] } ) => firstItem.type === type;
```

Comment:
`type => ( { item: [ firstItem ] } ) => firstItem.type === type;` => `type => item => item.type === type;` (since timeline just stores events as single objects now).